In [1]:
import numpy as np
import pandas as pd

# Transportation cost matrix (RM per unit)
cost = np.array([
    [4, 8, 8, 6, 10],   # Factory A
    [6, 4, 7, 3, 6],    # Factory B
    [8, 6, 9, 7, 5],    # Warehouse X
    [10, 8, 4, 6, 8]    # Warehouse Y
])

# Supply and Demand
supply = [60, 75, 50, 40]            # A, B, X, Y
demand = [40, 30, 50, 60, 45]        # C1 to C5

# Create allocation matrix filled with zeros
allocation = np.zeros((4, 5), dtype=int)

# Make copies for iteration
supply_copy = supply.copy()
demand_copy = demand.copy()

# Apply Northwest Corner Method
i = j = 0
while i < len(supply_copy) and j < len(demand_copy):
    alloc = min(supply_copy[i], demand_copy[j])
    allocation[i][j] = alloc
    supply_copy[i] -= alloc
    demand_copy[j] -= alloc
    if supply_copy[i] == 0:
        i += 1
    if demand_copy[j] == 0:
        j += 1

# Calculate total transportation cost
total_cost = np.sum(allocation * cost)

# Prepare allocation table for display
sources = ["Factory A", "Factory B", "Warehouse X", "Warehouse Y"]
customers = ["Customer 1", "Customer 2", "Customer 3", "Customer 4", "Customer 5"]
alloc_df = pd.DataFrame(allocation, index=sources, columns=customers)

# Display results
print("=== Allocation Table (Northwest Corner Method) ===")
print(alloc_df)
print("\nTotal Transportation Cost: RM", total_cost)


=== Allocation Table (Northwest Corner Method) ===
             Customer 1  Customer 2  Customer 3  Customer 4  Customer 5
Factory A            40          20           0           0           0
Factory B             0          10          50          15           0
Warehouse X           0           0           0          45           5
Warehouse Y           0           0           0           0          40

Total Transportation Cost: RM 1415


#Step By Step Explaination

#Step 1: Start at the Northwest (Top-Left) Corner
Begin at Factory A → Customer 1.

* Allocate the minimum of supply and demand.

* Subtract the allocated value from both supply and demand.

* Move right if supply is exhausted, down if demand is exhausted.

* Repeat until all supply and demand are satisfied.

#Step 2: Record Allocations.

#Step 3: Show the records in table.

#Step 4: Total Transportation Cost: RM 1415

#Intepretation

Using the Northwest Corner Method, the total transportation cost is RM 1415. This solution satisfies all customer demands using all available supply. However, this method doesn't consider costs while allocating, so while feasible, it's not guaranteed to be optimal. For cost optimization, advanced methods like Least Cost or MODI are preferred.
